### PR TITLE
ESP32 Improvements and mbedTLS support

### PR DESCRIPTION
**Pull request by Palantir555**
This PR has been cherry-picked-- Merge #183, #184 and #186 into development first.

[ Description ]

Summary: 

Better SDK integration to fix our connectivity issues:

- Our ESP32 BSP doesn't require its own `sdkconfig.h` anymore, in line with the design of the SDK's build system. This prevents conflicts in build configurations:
    - Moved the BSP sources into a custom subdirectory of `src/bsp/platform/esp32`, which excludes them from the `libxively.a` build
    - Created a `component.mk` file in the BSP directory. It will be included by the SDK to build the BSP as a shared library
    - The ESP32 demo's makefile tells the SDK where to find the BSP component
    - `xi_bsp_hton.h` now has forward declarations of the `lwip_hto*` functions instead of including `lwip/def.h`. That removes the last libxively.a dependency on `sdkconfig.h`
- Fixed usage of mbedTLS on the ESP32. Set `XI_BSP_TLS=mbedtls` when running make on libxively AND the application, and you'll get the SDK's hardware-accelerated version of mbedTLS
- Implement support for a Xively component in the IDF SDK. I've successfully integrated the library into the SDK, and our demo's Makefile doesn't misbehave when the component is available. Here's the SDK integration: https://github.com/Palantir555/esp-idf/tree/feature/add_xively_component/components/xively

Security/WolfSSL:

- Fixed the WolfSSL RNG_seed security warning for the ESP32 PRESET:
    - Create the macros `XI_PROVIDE_WOLFSSL_SEED_GENERATOR`  and `XI_PROVIDE_WOLFSSL_XTIME_XGMTIME` to move the WolfSSL callbacks from applications and platform BSPs into the WolfSSL BSP. Currently enabled for ESP32 only, but we can set them for the rest of the PRESETs and remove the common code from the examples

Other:

- `#if WOLFSSL_DEBUG_LOG`, call `wolfSSL_Debugging_ON();` from the WolfSSL BSP
- Fixed an issue with the ESP32's light show that turned the LED off shortly after being turned on from the internet.
- The ESP32's FW update logic to reboot/not reboot after a firmware update has been moved from the BSP to `main.c`

[ JIRA ]
@DellaBitta

[ Reviewer ]

[ QA ]
All moving parts (SDK, libxively, TLS library, BSP component, potential SDK integration, and the example) build and run, and I was able to connect to the broker using the latest version of the SDK. I've tested many configurations:

- libxively as an SDK component && manually-downloaded WolfSSL == OK
- libxively as an SDK component && SDK's mbedTLS == OK
- libxively's build system && ESP32 BSP as an SDK component && SDK's mbedTLS == OK
- libxively's build system && ESP32 BSP as an SDK component && WolfSSL == OK
- libxively as an SDK component is DISABLED && build the example from our examples directory == OK
- libxively as an SDK component is ENABLED && build the example from our examples directory == Error out and ask the user to disable the xively component
- libxively's build system && ESP32 BSP as an SDK component && our mbedTLS == Build errors

[ Release Notes ]
- Correctly integrate the ESP32 BSP with the manufacturer's SDK. Remove libxively.a's dependencies on `sdkconfig.h`, making `libxively.a` truly application-agnostic
- Implement mbedTLS support for the ESP32 preset. It uses the SDK's (hardware-accelerated) mbedTLS component
- Fix ESP32 demo light show bug
- Make WolfSSL's time and RNG callbacks available from our WolfSSL BSP.

### CODE REVIEW

File: examples/cc3220sf/xively_demo/.project [hunk 1]
```diff
@@ -28,7 +29,7 @@
 	<variableList>
 		<variable>
 			<name>XIVELY_LIBRARY_C_ROOT</name>
-			<value>file:/E:/dev/xively/xively-client-c</value>
+			<value>$%7BPARENT-3-PROJECT_LOC%7D</value>
```
Comment (Palantir555) [inline examples/cc3220sf/xively_demo/.project]: This is from the test I ran on the last 3220 PR. `$%7BPARENT-3-PROJECT_LOC%7D` is eclipse-talk for `../../../`

File: examples/esp32/xively_demo/main/main.c [hunk 2]
```diff
@@ -422,7 +422,26 @@ void esp32_xibsp_notify_chunk_written( size_t chunk_size, size_t offset )
     }
 }
 
-void esp32_xibsp_notify_update_applied()
+void esp32_xibsp_notify_update_applied( uint8_t app_updated )
```
Comment (Palantir555) [inline examples/esp32/xively_demo/main/main.c]: This is application-level logic (rebooting the device or not based on the package contents), so I've taken this opportunity to move it out of the BSP.

File: make/mt-config/mt-tls-mbedtls.mk [hunk 3]
```diff
@@ -8,3 +8,4 @@ XI_TLS_LIB_BIN_DIR ?= ./src/import/tls/mbedtls/library/
 XI_TLS_LIB_NAME ?= mbedtls mbedx509 mbedcrypto
 
 XI_CONFIG_FLAGS += -DXI_TLS_LIB_MBEDTLS
+XI_CONFIG_FLAGS += -DMBEDTLS_PLATFORM_MEMORY
```
Comment (Palantir555) [inline make/mt-config/mt-tls-mbedtls.mk]: Moved here from the BSP source

File: examples/esp32/xively_demo/main/main.c [hunk 4]
```diff
@@ -221,22 +226,12 @@ int8_t app_wifi_station_init( user_data_t* credentials )
...
 
-#if USE_HARDCODED_CREDENTIALS
-    wifi_config_t wifi_config =
-    {
-        .sta =
-        {
-            .ssid = USER_CONFIG_WIFI_SSID, .password = USER_CONFIG_WIFI_PWD
-        }
-    };
-#else
```
Comment (Palantir555) [inline examples/esp32/xively_demo/main/main.c]: This was redundant -- Hardcoded credentials are already in the *credentials struct

File: examples/cc3220sf/xively_demo/.project (hunk 1 above)
Comment (DellaBitta) [reply examples/cc3220sf/xively_demo/.project]: Eclipse ... you're a confusing IDE.

**Pull request merged**